### PR TITLE
fix encodeSQLValue for uint64 with high bit set

### PR DESCRIPTION
**Pull request by rck109d**
before this PR, calling `encodeSQLValue` with an `arg` of `uint64` (or `uint`, or pointers or aliases thereof) with a value greater than or equal to 1<<63 (high bit set) would result in an error from `driver.DefaultParameterConverter.ConvertValue` stating "uint64 values with high bit set are not supported" and then a call to `encodeSQLValue` would happen recursively forever.  This led to CPU throttling which led to overall service degradation (increased latencies and errors across all endpoints) for inventory2 in production earlier this week when very high timestamp values were attempted to be written to the DB.

These very large unsigned values might have arisen from negative integers incorrectly cast to unsigned integers; nevertheless, when such values are present on a field to be encoded for SQL, a non-terminating `encodeSQLValue` makes a bad situation much worse.

This PR adds many test cases to `TestEncodeSQL` and handles unsigned integers with their (64th) high bit set _before_ calling `driver.DefaultParameterConverter.ConvertValue`.

### CODE REVIEW

File: encode.go [hunk 1]
```diff
@@ -63,7 +63,16 @@ func encodeSQLValue(w io.Writer, arg interface{}) error {
 		// We may be in the presence of a type alias not supported by the
 		// database/driver DefaultParameterConverter. Special handling.
 		value := reflect.ValueOf(arg)
-		if baseKinds[value.Kind()] {
+		kind := value.Kind()
+		// try uint64 with high bit set first to avoid infinite recursion.
+		// 9223372036854775808 is the lowest uint64 with high bit set.
+		if kind == reflect.Uint64 && value.Uint() >= 9223372036854775808 {
```
Comment (ryho) [inline encode.go:69]: You should use `>= 1<<63` so that you can you can see at a glance that this is correct.

**Review by ryho — APPROVED** [review]

File: encode.go (hunk 1 above)
Comment (danrice-square) [reply encode.go]: Maybe uint(1) << 63?

Comment (rck109d) [reply encode.go]: fixed

Comment (rck109d) [reply encode.go]: going to circle back to handle both `uint64` and `uint` and their pointers

Comment (rck109d) [reply encode.go]: ok, pushed once more—maybe for the last time

File: encode.go [hunk 2]
```diff
@@ -64,7 +80,11 @@ func encodeSQLValue(w io.Writer, arg interface{}) error {
 		// database/driver DefaultParameterConverter. Special handling.
 		value := reflect.ValueOf(arg)
 		if baseKinds[value.Kind()] {
-			return encodeSQLValue(w, asKind(value))
+			asKind := asKind(value)
+			if reflect.DeepEqual(arg, asKind) { // avoid infinite recursion with same arg
+				return fmt.Errorf("unsupported type %T: ConvertValue error: %w", arg, err)
+			}
```
Comment (rck109d) [inline encode.go:86]: this piece would have stopped the infinite recursion which prompted this PR (can see it for oneself by removing the above-added preemptive encoding and running `TestEncodeSQL`)

File: encode.go (hunk 1 above)
Comment (rck109d) [reply encode.go]: yes, that's pretty bad, `defaultConverter.ConvertValue` does not behave consistently for high-bit-set `uint64` versus `uint`.  And if they "fixed" that then the `uint` flavor would start erroring too.
this issue would be protected against now that this PR's special casing has been moved above the default converter.

File: encode.go (hunk 2 above)
Comment (rck109d) [reply encode.go:87]: none of the test cases in `TestEncodeSQL` even trigger a single recursive call.
I'm now very dubious of the comment
```go
// We may be in the presence of a type alias not supported by the
// database/driver DefaultParameterConverter. Special handling.
```
and I suspect that the recursive call might only apply for (*)uint(64) cases

Comment (rck109d) [reply encode.go:87]: now that the (*)uint(64) encoding is special-cased before calling `driver.DefaultParameterConverter.ConvertValue`
I'm no longer able even to provoke a recursive call to `encodeSQLValue` even with many type alias test case additions.  That makes me very skeptical of the comment
```go
// We may be in the presence of a type alias not supported by the
// database/driver DefaultParameterConverter. Special handling.
```

**Review by danrice-square — APPROVED** [review]

**Pull request merged**